### PR TITLE
More protocols are introduced

### PR DESCRIPTION
**Pull request by artemstepanenko**

### CODE REVIEW

File: PetrMazepa/Base.lproj/SearchedArticleCell.xib [hunk 1]
```diff
@@ -1,8 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="9531" systemVersion="14F27" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES">
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="10116" systemVersion="15E65" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES">
     <dependencies>
-        <deployment identifier="iOS"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="9529"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="10085"/>
```
Comment (kapaljaka) [inline PetrMazepa/Base.lproj/SearchedArticleCell.xib]: I think, we don't need this changes.

File: PetrMazepa/Source/ArticleDetailsModule.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+//  ArticleDetailsModule.swift
+//  PetrMazepa
+//
+//  Created by Artem Stepanenko on 17/06/16.
+//  Copyright © 2016 TramCoders. All rights reserved.
+//
+
+import UIKit
+
+protocol IArticleDetailsView: class {
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticleDetailsModule.swift]: Why "I"?

File: PetrMazepa/Source/ArticleDetailsModule.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    
+    func reloadImage()
+    func reloadHtmlText()
+    func updateFavouriteState(favourite: Bool)
+    func updateBarsVisibility(visible: Bool)
+    func showError(error: NSError?)
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticleDetailsModule.swift]: maybe use swift Error here?

File: PetrMazepa/Source/ArticleDetailsModule.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    func applicationWillResignActive()
+    func backTapped()
+    func favouriteTapped()
+    func shareTapped()
+    func closeActionTapped()
+    func retryActionTapped()
+    func scrollViewWillBeginDragging(offset offset: CGFloat)
+    func scrollViewDidScroll(offset offset: CGFloat, contentHeight: CGFloat)
+    func textDidLoad()
+}
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticleDetailsModule.swift]: What the point if this protocol? Does it have any default implementation? Why it has so many methods?

File: PetrMazepa/Source/ArticleDetailsViewController.swift [hunk 5]
```diff
@@ -173,52 +164,46 @@ class ArticleDetailsViewController: UIViewController, UICollectionViewDataSource
...
-            alertController.addAction(retryAction)
-            
-            self.presentViewController(alertController, animated: true, completion: nil)
-        }
+ 
+    func showError(error: NSError?) {
+        
+        let alertController = UIAlertController(title: nil, message: "При получении статьи произошла ошибка", preferredStyle: .Alert)
+        
+        let closeAction = UIAlertAction(title: "Закрыть", style: .Cancel, handler: { _ in
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticleDetailsViewController.swift]: It's more convenient to use "Cancel" -> "Отмена" here I think?

File: PetrMazepa/Source/ArticleDetailsViewModel.swift [hunk 6]
```diff
@@ -8,19 +8,14 @@
...
 
-class ArticleDetailsViewModel : ViewModel {
+class ArticleDetailsViewModel : ViewModel, IArticleDetailsViewModel {
     
-    var imageLoaded: ((image: UIImage?) -> Void)?
-    var textLoaded: ((htmlText: String?) -> Void)?
-    var favouriteStateChanged: ((favourite: Bool) -> Void)?
-    var barsVisibilityChanged: ((visible: Bool) -> Void)?
-    var errorOccurred: ((error: NSError?) -> Void)?
+    var view: IArticleDetailsView?
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticleDetailsViewModel.swift]: Why optional? ViewModel can exist without View?

File: PetrMazepa/Source/ArticlesModule.swift [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    func searchTapped()
+    func refreshTriggered()
+    func lastReadArticleTapped()
+    func switchOffActionTapped()
+    func cancelActionTapped()
+    func retryActionTapped()
+    func articleTapped(index index: Int)
+    func articleModel(index index: Int) -> ArticleCellModel
+    func willBeginDragging(offset offset: CGFloat)
+    func didScroll(contentOffset contentOffset: CGFloat, distanceToBottom: CGFloat)
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesModule.swift]: Again. What is the point of such big-ass protocols ? Where is the fucking composition here, man?

File: PetrMazepa/Source/ArticlesViewController.swift [hunk 8]
```diff
@@ -137,136 +122,117 @@ class ArticlesViewController: UIViewController, UICollectionViewDelegate, UIColl
...
 
-            self.layout.deleteAllArticles()
-            self.layout.insertArticles(newCount)
-            self.collectionView.reloadData()
-        }
+extension ArticlesViewController: IArticlesView {
+    
+    func loadingInOfflineModeFailed() {
+        // message
+        let message = NSLocalizedString("ArticlesOfflineLoadingFailed_message", comment: "")
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewController.swift]: afaiu we use alert controller with two buttons several times in our app. So how about create an extension on alertController to remove all this boilerplate stuff?

File: PetrMazepa/Source/ArticlesViewController.swift [hunk 9]
```diff
@@ -137,136 +122,117 @@ class ArticlesViewController: UIViewController, UICollectionViewDelegate, UIColl
...
+            self.view.layoutIfNeeded()
         }
     }
     
-    private func noArticlesVisibleChangedHandler() -> ((visible: Bool) -> Void) {
-        return { visible in
-            
-            self.collectionView.hidden = visible
-            self.noArticlesView.hidden = !visible
+    func articlesInserted(range range: Range<Int>) {
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewController.swift]: articlesInserted(inRange...

File: PetrMazepa/Source/ArticlesViewController.swift [hunk 10]
```diff
@@ -137,136 +122,117 @@ class ArticlesViewController: UIViewController, UICollectionViewDelegate, UIColl
...
-            // TODO:
-        }
+    func navigationBarVisibilityChanged(visible visible: Bool, animated: Bool) {
+        self.navigationController?.setNavigationBarHidden(!visible, animated: animated)
     }
     
-    private func loadingMoreStateChangedHandler() -> ((loadingMore: Bool) -> Void) {
-        return { loadingMore in
-            // TODO:
+    func lastReadArticleVisibilityChanged(visible visible: Bool, animated: Bool) {
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewController.swift]: lastReadArticleVisibilityChanged(to ...

File: PetrMazepa/Source/ArticlesViewController.swift [hunk 11]
```diff
@@ -137,136 +122,117 @@ class ArticlesViewController: UIViewController, UICollectionViewDelegate, UIColl
...
-            self.layout.insertArticles(range.count)
-            let insertedIndexPaths = self.indexPaths(range: range)
-            
-            if range.startIndex == 0 {
-                self.collectionView.reloadData()
-            } else {
-                self.collectionView.insertItemsAtIndexPaths(insertedIndexPaths)
-            }
-        }
+    private func indexPaths(range range: Range<Int>) -> [NSIndexPath] {
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewController.swift]: indexPaths(fromRange ...

File: PetrMazepa/Source/ArticlesViewController.swift [hunk 12]
```diff
@@ -137,136 +122,117 @@ class ArticlesViewController: UIViewController, UICollectionViewDelegate, UIColl
...
-            let cancelAction = UIAlertAction(title: cancel, style: .Default, handler: { _ in
-                self.model.cancelActionTapped()
-            })
-            
-            alertController.addAction(cancelAction)
-            alertController.addAction(retryAction)
-            
-            self.presentViewController(alertController, animated: true, completion: nil)
-        }
+    func refreshingStateChanged(refreshing refreshing: Bool) {
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewController.swift]: func refreshingStateChanged(to..
there is wwdc session about syntax of the swift 3.0

File: PetrMazepa/Source/ArticlesViewModel.swift [hunk 13]
```diff
@@ -8,26 +8,15 @@
...
-    var allArticlesDeleted: (() -> Void)?
-    var articlesUpdated: ((newCount: Int) -> Void)?
-    var loadingInOfflineModeFailed: (() -> Void)?
-    var errorOccurred: (() -> Void)?
-    var refreshingStateChanged: ((refreshing: Bool) -> Void)?
-    var loadingMoreStateChanged: ((loadingMore: Bool) -> Void)?
-    var noArticlesVisibleChanged: ((visible: Bool) -> Void)?
-    var lastReadArticleVisibleChanged: ((visible: Bool, animated: Bool) -> Void)?
-    var navigationBarVisibleChanged: ((visible: Bool, animated: Bool) -> Void)?
+    private var view: IArticlesView?
```
Comment (kapaljaka) [inline PetrMazepa/Source/ArticlesViewModel.swift]: Can be optional ?

File: PetrMazepa/Source/PersistentImageStorage.swift [hunk 14]
```diff
@@ -19,15 +21,17 @@ class PersistentImageStorage {
 
     private let fileManager: NSFileManager
 
-    init() {
+    init(tracker: ITracker) {
```
Comment (kapaljaka) [inline PetrMazepa/Source/PersistentImageStorage.swift]: I don't like an idea of Tracker being an argument to construct ImageStorage. Looks weird to me. You expect something object cannot function properly without as an argument. And Tracker is something supplementary and optional. What do you think?

File: PetrMazepa/Source/Router.swift [hunk 15]
```diff
@@ -119,32 +119,32 @@ class ScreenFlow: NSObject, ArticleDetailsPresenter, SettingsPresenter, SearchPr
...
                 return
             }
             
-            Tracker.trackShare(article, activityType: activityType)
+            self?.tracker.trackShare(article, activityType: activityType)
         }
     }
     
-    private func createArticlesViewModel() -> ArticlesViewModel {
+    private func createArticlesViewModel(view view: IArticlesView) -> ArticlesViewModel {
```
Comment (kapaljaka) [inline PetrMazepa/Source/Router.swift]: What would you say about renaming this whole bunch of methods to something like
private func createViewModel(forView...
private func createViewModel(forArticleDetails...
private func createViewModel(forSettings...
etc?

File: PetrMazepa/Source/SearchViewController.swift [hunk 16]
```diff
@@ -8,7 +8,9 @@
 
 import UIKit
 
-class SearchViewController: UIViewController, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate {
+class SearchViewController: UIViewController, ISearchView, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate {
```
Comment (kapaljaka) [inline PetrMazepa/Source/SearchViewController.swift]: Hella a lot of view stuff this view controller conforms to. Maybe create separate reusable classes for at least tableView(collection view, whatever view) datasource and delegate?

File: PetrMazepa/Source/Tracker.swift [hunk 17]
```diff
@@ -9,7 +9,7 @@
 import UIKit
 import Crashlytics
 
-class Tracker: NSObject {
+class Tracker: ITracker {
```
Comment (kapaljaka) [inline PetrMazepa/Source/Tracker.swift]: add final for all the classes that are not intended to be subclassed

File: PetrMazepaTests/SettingsViewModelSpec.swift [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+//        
+//        describe("a load method") {
+//            context("nil") {
+//                it("nil is nil") {
+//                    expect(nil).to(beNil())
+//                }
+//            }
+//        }
+//    }
+//}
```
Comment (kapaljaka) [inline PetrMazepaTests/SettingsViewModelSpec.swift]: fuck tests ))

**Comment by kapaljaka** [comment]
Overall I don't get the point about creating whole bunch of I-prefixed protocols that are just replicate old interfaces.  What for?

File: PetrMazepa/Source/ArticleDetailsViewController.swift (hunk 5 above)
Comment (artemstepanenko) [inline PetrMazepa/Source/ArticleDetailsViewController.swift]: Done.

File: PetrMazepa/Source/ArticlesViewController.swift (hunk 9 above)
Comment (artemstepanenko) [inline PetrMazepa/Source/ArticlesViewController.swift]: Done.

File: PetrMazepa/Source/ArticlesViewController.swift (hunk 10 above)
Comment (artemstepanenko) [inline PetrMazepa/Source/ArticlesViewController.swift]: Done.

File: PetrMazepa/Source/ArticlesViewController.swift (hunk 11 above)
Comment (artemstepanenko) [inline PetrMazepa/Source/ArticlesViewController.swift]: Done.

File: PetrMazepa/Source/ArticlesViewController.swift (hunk 12 above)
Comment (artemstepanenko) [inline PetrMazepa/Source/ArticlesViewController.swift]: Done.

**Comment by artemstepanenko** [comment]
Is there anything unclear after the discussion we had in offline? Please, take a look into SettingsViewModelSpec, it's very basic implementation of unit tests I wanted to achieve.

File: PetrMazepa/Base.lproj/SearchedArticleCell.xib [hunk 19]
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="9531" systemVersion="14F27" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES">
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="10116" systemVersion="15E65" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES">
```
Comment (rkopaliani) [inline PetrMazepa/Base.lproj/SearchedArticleCell.xib]: what about this? reset if not needed

File: PetrMazepa/Source/ArticleDetailsViewController.swift [hunk 20]
```diff
@@ -173,52 +164,46 @@ class ArticleDetailsViewController: UIViewController, UICollectionViewDataSource
...
-            
-            alertController.addAction(closeAction)
-            alertController.addAction(retryAction)
-            
-            self.presentViewController(alertController, animated: true, completion: nil)
-        }
+ 
+    func showError(error: NSError?) {
+        
+        let alertController = UIAlertController(title: nil, message: "При получении статьи произошла ошибка", preferredStyle: .Alert)
```
Comment (rkopaliani) [inline PetrMazepa/Source/ArticleDetailsViewController.swift]: what about creating alert controller extension?

File: PetrMazepa/Source/ArticleDetailsViewModel.swift [hunk 21]
```diff
@@ -8,19 +8,14 @@
 
 import UIKit
 
-class ArticleDetailsViewModel : ViewModel {
+class ArticleDetailsViewModel : ViewModel, ArticleDetailsViewModelProtocol {
```
Comment (rkopaliani) [inline PetrMazepa/Source/ArticleDetailsViewModel.swift]: add final for all viewModel classes, they are not to be subclassed afaik

**Pull request merged**